### PR TITLE
[flutter_tools] boolArg refactor

### PR DESCRIPTION
**Pull request by amanv8060**
Deprecate the boolArg function, machine replace its usages, make new function handling nulls bool? boolArg

part of #101595

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 1]
```diff
@@ -1519,7 +1519,11 @@ abstract class FlutterCommand extends Command<void> {
   ApplicationPackageFactory? applicationPackages;
 
   /// Gets the parsed command-line option named [name] as a `bool`.
-  bool boolArg(String name) => argResults?[name] as bool? ?? false;
+  @Deprecated('Deprecated in favor of boolArg')
```
Comment (amanv8060) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1522]: @christopherfujino do we need to mark it deprecated as per the [policy](https://github.com/flutter/flutter/wiki/Tree-hygiene#deprecations)?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1522]: I asked @Jasguerrero to review this, as it was his proposal.

Comment (Jasguerrero) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1522]: > @christopherfujino do we need to mark it deprecated as per the [policy](https://github.com/flutter/flutter/wiki/Tree-hygiene#deprecations)?

@amanv8060 Unlike the framework the tool is not imported by customers so no need to follow the wiki's policy in this case.

Comment (amanv8060) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1522]: Sure, I will remove it.

Comment (amanv8060) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1522]: @Jasguerrero, will this be a test-exempt?

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 2]
```diff
@@ -1519,7 +1519,11 @@ abstract class FlutterCommand extends Command<void> {
   ApplicationPackageFactory? applicationPackages;
 
   /// Gets the parsed command-line option named [name] as a `bool`.
-  bool boolArg(String name) => argResults?[name] as bool? ?? false;
+  /// This has been deprecated, use [boolArg] instead.
+  bool boolArgDeprecated(String name) => argResults?[name] as bool? ?? false;
+
+  /// Gets the parsed command-line option named [name] as a `bool?`.
+  bool? boolArg(String name) => argResults?[name] as bool?;
```
Comment (Jasguerrero) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1526]: This function still raises the same ArgumentError as the deprecated one (when the "name" is not present on argResults) I suggest updating this function to handle that case. I am not sure if you can do it with a ternary operator in one line. In the proposal of the issue is an example of how to do it.

I also suggest adding 3 new tests under https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/args_test.dart

- argResult is null (boolArg should return null)
- "name" not present in argResult (boolArg? should return null)
- "name" present in argResult (boolArg return true/false depending on value)

**Review by Jasguerrero — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/runner/flutter_command.dart (hunk 2 above)
Comment (amanv8060) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1526]: I will attempt to write a test for this next week.

Comment (Jasguerrero) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:1526]: @amanv8060 Here is an example you can use to write your tests https://github.com/flutter/flutter/pull/103231

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 3]
```diff
@@ -1519,7 +1519,16 @@ abstract class FlutterCommand extends Command<void> {
...
-  bool boolArg(String name) => argResults?[name] as bool? ?? false;
+  /// This has been deprecated, use [boolArgDeprecated] instead.
+  bool boolArgDeprecated(String name) => argResults?[name] as bool? ?? false;
+
+  /// Gets the parsed command-line option named [name] as a `bool?`.
+  bool? boolArg(String name) {
+    if (argResults == null || !argParser.options.containsKey(name)) {
+      return null;
+    }
+    return argResults?[name] as bool?;
```
Comment (Jasguerrero) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1530]: `return  argResults?[name] as bool? ?? false`

File: packages/flutter_tools/test/general.shard/args_test.dart [hunk 4]
```diff
@@ -22,6 +24,24 @@ void main() {
...
   }));
+
+  testUsingContext('bool? safe argResults', () async {
+    final DummyFlutterCommand command = DummyFlutterCommand(
+        commandFunction: () async {
+          return const FlutterCommandResult(ExitStatus.success);
+        }
+    );
+    final FlutterCommandRunner runner = FlutterCommandRunner(verboseHelp: true);
+    command.argParser.addFlag('key');
```
Comment (Jasguerrero) [inline packages/flutter_tools/test/general.shard/args_test.dart:35]: Let's add 2 more expected

```
command.argParser.addFlag('key-false');
await runner.run(<String>['dummy', '--key']);
expect(command.boolArg('key-false'), false);
expect(command.boolArgDeprecated('key-false'), false);
```

I believe that with your current code key-false will map to null in boolArg without my previous comment

**Review by Jasguerrero — CHANGES_REQUESTED** [review]

**Review by Jasguerrero — APPROVED** [review]
LGTM

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**